src/features/APIWithHooks.ts

```diff
@@ -0,0 +1,61 @@
+import {IAPI} from '../APICore';
+
+export interface IAPIHooks {
+    beforeAnyRequest?: (url: string, args: RequestInit) => void;
+    afterAnySuccess?: <T>(url: string, args: RequestInit, response: T) => T;
+    afterAnyException?: (url: string, args: RequestInit, exception: Error) => boolean;
+}
+
+export class APIWithHooks<TAPI extends IAPI = IAPI> implements IAPI {
```

Comment:
I think `APIWithHooks` should extend `API`. In API we already have a `request` method that is called for every requests. We could add the before/after handlers and it would be enough?

Reply:
I didn't want to expose any inner working of `API`, I would have to change `request` to a protected method, which would allow extenders to override the method.

So I leaned towards a composing wrapper instead of an inheriting one

Reply:
The more I think about it the more I'm leaning toward the extend instead of the composition. Here I feel like you reimplement everything which expose more of the API class than just making the request method protected

Reply:
So I tried it, and I think it broke things too hard from the original design.

Here is the implementation with the extends:

```
export class APIWithHooks<TAPI extends IAPI = IAPI> extends API {
    constructor(config: APIConfiguration, private hooks: IAPIHooks) {
        super(config);
    }

    protected async request<T>(route: string, args: RequestInit): Promise<T> {
        this.hooks.beforeAnyRequest?.(route, args);

        try {
            const response = await super.request<T>(route, args);;
            this.hooks.afterAnySuccess?.<T>(route, args, response);
            return response;
        } catch (exception) {
            if (!this.hooks.afterAnyException?.(route, args, exception)) {
                throw exception;
            }
        }
    }
}
```

but this is not the same as what I had before. Now, instead of creating an `API` instance, I have to create an `APIWithHooks` instance. It would be fine if I wanted to replace some of the APIs with this new implementation, but here is the main issue:

I can't use this to "enhance" the API with features. My original `type IAPIFeature = (api: IAPI) => IAPI;` would have to be changed to `type IAPIFeature = (config: IAPI) => IAPI;
` , but now we completely lose the ability to "chain" features together.

I might be missing something here, but I do think the composition design is what we need to properly support this feature set